### PR TITLE
net/ieee802154: Use `eui64_t` and `network_uint16_t` in `ieee802154_t` for addresses

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

- Updated `ieee802154_t` to use
    - `eui64_t` for storing the long address
    - `network_uint16_t` for storing the short address
- Updated the shared IEEE 802.15.4 code to that API
- Updated `at86rf2xx` to that API
- Updated `cc2420` to that API
- Fixed a super strange bug in `cc2420`
- Updated `kw2xrf` to that API
- Updated `mrf24j40` to that API
- Updated `native` to that API
- Updated `nrf52/nrf802154` to that API
- Updated `xbee` to that API
- Updated test for `at86rf2xx` to that API

### Testing procedure

`ping6` should still work on:

- [ ] `at86rf2xx`
- [ ] `cc2420`
- [ ] `kw2xrf`
- [ ] `mrf24j40`
- [ ] `native`
- [ ] `nrf52/nrf802154`
- [ ] `xbee`

### Issues/PRs references

Found this as useful in https://github.com/RIOT-OS/RIOT/pull/12600

### CODE REVIEW

**Comment by maribu** [comment]
I think it's fair to say this is an internal API change, so that no deprecation is needed. But I think it has impact enough to require 2 ACKs

**Comment by kaspar030** [comment]
netdev intentionally doesn't use "fancy" representations for basic values. That does pile up dependencies, which might not be wanted for generic interfaces.

I also assumed that the represesentation of eui64_t forces alignment on 8 bytes (due to the uint64_t member), but I think that's not happening here.

**Comment by maribu** [comment]
> I also assumed that the represesentation of eui64_t forces alignment on 8 bytes

On 4 bytes on 32 bit platforms. Which is a good point for doing so. People might otherwise just cast the byte array to an `eui64_t` and use C assignments, which could then be unaligned.

I'd say either we should use `eui64_t` consistently everywhere, or not at all.

**Comment by maribu** [comment]
> netdev intentionally doesn't use "fancy" representations for basic values. That does pile up dependencies, which might not be wanted for generic interfaces.

I thought about this a bit and I cannot follow your argumentation.

1. It doesn't add more dependencies
     - The types only need to `#include` a header, so if an additional dependency would be added, adapting the code would need additional `#include`s. But it didn't. The IEEE802154 drivers do already depend on this anyway
2. There seems to be no relevant difference in the context
    - The `netdev_ieee802154` (which is not the generic `netdev`, but a "subclass" in OOP terminology) is seeing the address as EUI-64, not as opaque byte string. (In fact e.g. the `at86rf2xx` driver is using `eui64_t` already when requesting the address from LUID)
    - To me, this seems to be pretty much the same context as where `eui64_t` is used (leaving out context details not relevant to the use of `eui64_t`)
    - So every argument for and against the use of the `eui64_t` should apply in the same way to the use case in the netdev drivers specific to 802.15.4. So if one considers the use of `eui64_t` to be valuable in one case, one should come to the same conclusion in the other. (Unless I'm overlooking some difference in the use cases that is relevant here.)

**Comment by maribu** [comment]
Is there btw any reason that `eui64_t` is not just `typedef`ed as [`network_64_t`](http://api.riot-os.org/byteorder_8h.html#ad85448169c2299064ee17b507fba692f)?

Now for `eui64_t e` I could use `e.uint8` or `e.uint64.u8` or `e.uin16.u8` to yield the same result. And to get the 64 bit value, I would have to use `e.uint64.u64`. This seems really clumsy.

**Comment by miri64** [comment]
> Is there btw any reason that `eui64_t` is not just `typedef`ed as [`network_64_t`](http://api.riot-os.org/byteorder_8h.html#ad85448169c2299064ee17b507fba692f)?

Because sometimes you need the array representation for it.

Regarding the dependencies: I think it makes no sense to not use a header that helps to keep an address type correct, just to keep a module to itself. This encourages code duplication, and (as we see in this instances) copied errors.

File: cpu/native/socket_zep/socket_zep.c [hunk 1]
```diff
@@ -383,16 +383,16 @@ void socket_zep_setup(socket_zep_t *dev, const socket_zep_params_t *params)
...
     for (unsigned i = 0; i < ss_len; i++) { /* generate NIC from local source */
         unsigned addr_idx = (i % (IEEE802154_LONG_ADDRESS_LEN / 2)) +
                             (IEEE802154_LONG_ADDRESS_LEN / 2);
-        dev->netdev.long_addr[addr_idx] ^= ss_array[i];
+        dev->netdev.long_addr.uint8[addr_idx] ^= ss_array[i];
     }
-    dev->netdev.short_addr[0] = dev->netdev.long_addr[6];
-    dev->netdev.short_addr[1] = dev->netdev.long_addr[7];
+    dev->netdev.short_addr.u8[0] = dev->netdev.long_addr.uint8[6];
+    dev->netdev.short_addr.u8[1] = dev->netdev.long_addr.uint8[7];
```
Comment (kaspar030) [inline cpu/native/socket_zep/socket_zep.c]: But this is supposed to be an improvement?

**Comment by maribu** [comment]
> > Is there btw any reason that `eui64_t` is not just `typedef`ed as [`network_64_t`](http://api.riot-os.org/byteorder_8h.html#ad85448169c2299064ee17b507fba692f)?
> 
> Because sometimes you need the array representation for it.

But isn't `n.u8` for `network_64_t n` also a way to access the array representation?

File: cpu/native/socket_zep/socket_zep.c (hunk 1 above)
Comment (maribu) [reply cpu/native/socket_zep/socket_zep.c]: This argument applies to the use of `eui64_t` in general, not to using it consistently.

But in fact the use of `eui64_t` will allow making this a bit easier, I will add a fixup to do this.

File: drivers/at86rf2xx/at86rf2xx_getset.c [hunk 2]
```diff
@@ -132,41 +132,34 @@ static const uint8_t dbm_to_rx_sens[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
...
 uint64_t at86rf2xx_get_addr_long(const at86rf2xx_t *dev)
 {
-    uint64_t addr;
-    uint8_t *ap = (uint8_t *)(&addr);
-
-    for (int i = 0; i < 8; i++) {
-        ap[i] = dev->netdev.long_addr[i];
-    }
-    return addr;
+    return dev->netdev.long_addr.uint64.u64;
```
Comment (kaspar030) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: this certainly is, but the original code was easy to improve on.

**Comment by miri64** [comment]
> > > Is there btw any reason that `eui64_t` is not just `typedef`ed as [`network_64_t`](http://api.riot-os.org/byteorder_8h.html#ad85448169c2299064ee17b507fba692f)?
> > 
> > 
> > Because sometimes you need the array representation for it.
> 
> But isn't `n.u8` for `network_64_t n` also a way to access the array representation?

I guess not pulling in the `byteorder.h` header was the goal here ;-)

**Comment by maribu** [comment]
> > 
> 
> I guess not pulling in the `byteorder.h` header was the goal here ;-)

It uses `network_uint64_t` and `network_uint16_t`, which both pull in `byteorder.h`. So it is already pulled in.

Let me rephrase the question: To me changing the definition of `eui64_t` to be an alias to `network_uint64_t` instead of having it `network_uint64_t` in a `union` seems beneficial. Do I miss anything?

If not, I'd like to create a PR that would just apply this change.

**Comment by maribu** [comment]
@kaspar030: I think I didn't properly express the goals of this PR.

The goal is to consistently make use of `eui64_t`; At some points the code gets (accidentially) more readable, at others it gets slightly less readable. But readability of the code is not a goal of this PR.

And this are the reasons why I believe consistent use can be beneficial:

- Other APIs like [`void luid_get_eui64(eui64_t *addr )`](http://api.riot-os.org/group__sys__luid.html#gab4a2825933935633f5f681abd9f84b4c) already expect this type
    - Passing byte arrays as argument is not only clumsy (as it requires explicit casts), but also has potential issues with unaligned memory access. (Currently in `netdev_ieee802154_t` the alignment happens to be correct by chance, if I see this correctly. But changes to the struct could prove fatal, if the alignment requirement is not properly enforced.)
- The types can avoid confusion about byteorder, as the type must only used store data in network byte order
    - Just changing APIs to using this type let to https://github.com/RIOT-OS/RIOT/pull/12606 fixing a byteorder confusion bug in the `at86rfxxx` driver
- consistency by itself as inherit merit

**Comment by maribu** [comment]
May I squash?

**Comment by miri64** [comment]
> > I guess not pulling in the `byteorder.h` header was the goal here ;-)
> 
> It uses `network_uint64_t` and `network_uint16_t`, which both pull in `byteorder.h`. So it is already pulled in.
> 
> Let me rephrase the question: To me changing the definition of `eui64_t` to be an alias to `network_uint64_t` instead of having it `network_uint64_t` in a `union` seems beneficial. Do I miss anything?
> 
> If not, I'd like to create a PR that would just apply this change.

If its not the dependency then to my knowledge there is no specific reason (except maybe that some of the resolutions provided by `network_uint64_t` are not necessary for `eui64_t`.

**Comment by benpicco** [comment]
What's the verdict?
I think an added dependency on a header file should not be a blocker here, this is a good cleanup.

A rebase is needed btw.

**Comment by maribu** [comment]
Rebased to solve the merge conflict. (And as this quite significantly changed the history anyway, I squashed as well.)

**Comment by maribu** [comment]
Let me point out that this change can prevent two kinds of bugs:

- Misaligned access to the address, as other APIs already start expecting data to be aligned as `eui64_t`
- Endianess confusion. (The `eui64_t` is defined to always be in network byte order.)

The consistent use of the same type for the same thing is nice to have, but maybe not that convincing as an argument. But the argument that this can help to prevent bugs is pretty convincing to me.

**Comment by benpicco** [comment]
Looks like `at86rf2xx` still needs an update.

**Comment by kaspar030** [comment]
> * Misaligned access to the address, as other APIs already start expecting data to be aligned as `eui64_t`

Hm, I don't quite agree here. Defining "eui64_t" as uint64_t is causing the alignment issues...

> * Endianess confusion. (The `eui64_t` is defined to always be in network byte order.)

Same. `uint8_t[8]` has perfectly clear byte order.

IMO defining eui64_t as it is now causes problems (and increases complexity) for the minimal convenience of having access to *parts* of the eui64 using different types. The price is 1) alignment issues, 2) endianness confusion, 3) dependency on the RIOT specific, cpu specific byteorder.h...
These are the kind of issues that personally I'd like to avoid when designing a portable API and/or network stack.

As seen here, almost all accesses use `uint8` anyways.

I'd be much more in favour of adding generic static inline functions like `eui64_compare(a, b)` (calling memcmp). And turning the short address into `uint8_t[2]`.

**Comment by kaspar030** [comment]
> I'd be much more in favour of adding generic static inline functions like `eui64_compare(a, b)` (calling memcmp).

... and changing eui64_t to a struct only containing the `uint8_t` array, to avoid alignment and byteorder issues.

**Comment by maribu** [comment]
The EUI-64 is sadly defined as an unsigned 64 bit number by IEEE. And e.g. the AT86RF2xx driver family is storing this number in little endian, not in network byte order. The endianess issues are real (and a bug in the AT86RF2xx driver is proving this).

**Comment by maribu** [comment]
> and changing eui64_t to a struct only containing the `uint8_t` array

I personally have nothing against this. Would your decision on this PR be influenced by this? The issue with the alignment would be solved even without using `eui64_t` consistently, but still the byte order issue would be an argument for also using the type here. (And consistentcy by itself has some inherent merit.)

**Comment by fjmolinas** [comment]
ping @kaspar030

**Comment by fjmolinas** [comment]
Lets try to unblock this one, the this PR seems to be blocked on the choice between:

1. changing `eui64_t` to a struct only containing the `uint8_t` array
1. changing the type of the member in `netdev_ieee802154_t::long_addr`

Both fix memory alignment issues. Endianess will still need to be dealt with in either case. But in any case, the current solution [1] at least improves on the situation (with maybe a little lost readability).

@kaspar030 you expressed you didn't like option [1] but @maribu seemed in favor of adapting to [2] if that would unblock the PR, seems we are all willing to give some concessions, so lets reach a decision :)

**Comment by kaspar030** [comment]
>     1. changing `eui64_t` to a struct only containing the `uint8_t` array

That would be fine with me, if that removes the byteorder.h dependency from eui64.h.

**Comment by miri64** [comment]
This should be migrated carefully though. The NDP code also uses `eui64.h`. There might be some trip-wires when migrating this code.

**Comment by fjmolinas** [comment]
This one should probably wait till after feature freeze, but it would be nice to get it in soon!

**Comment by fjmolinas** [comment]
@maribu are you still interested in this one?

**Comment by maribu** [comment]
> @maribu are you still interested in this one?

Well, we need to fix the memory alignment issue one way or the other. This would be my preferred way.

**Comment by miri64** [comment]
Reason for closure?

**Comment by maribu** [comment]
This morning I just removed all stale branches. Seems like I picked two which are not that stale, after all :-)

**Comment by maribu** [comment]
But anyway, with https://github.com/RIOT-OS/RIOT/pull/14611 being blocked, I could have kept this closed, I guess.
